Minimal integration of profiling api

bluenote10 opened the pull request:
Following up on https://github.com/godot-rust/book/issues/5, this is a naive way to integrate the profiling API, but it seems to work:

```rust
// Usage:
godot::nativescript::utils::profiling_add_data("::100::foo", 1000);
```

![image](https://user-images.githubusercontent.com/3620703/91095098-1318e600-e65c-11ea-9519-b5f30d948802.png)

Note: I think the API JSON actually has a small bug:

```json
          {
            "name": "godot_nativescript_profiling_add_data",
            "return_type": "void",
            "arguments": [
              ["const char *", "p_signature"],
              ["uint64_t", "p_line"]
            ]
          }
```
Note that the second argument says `p_line`, but it really is the time in microseconds. I think the name confusion stems from its usage example [here](https://github.com/godotengine/godot-cpp/blob/master/include/core/GodotProfiling.hpp).

toasteater commented:
Thanks for the PR! I agree that a high-level wrapper for `godot_nativescript_profiling_add_data` is a good idea, and following this we might want to add a procedural macro to automate the timing process, too. Unfortunately, while I understand that this is your first code contribution to this repo, and don't mean to sound harsh in any way, I don't think the current code in this PR is an example of a good high-level API:

1. The `signature` argument is not valid-by-construction. If the C API expects a specific format, the idiomatic way is to take a custom struct and format it into the correct string in the function body. I do recognize that this takes a bit of time, but it shouldn't affect the measurement since it's done outside the function. One would probably want to disable profiling for release builds anyway.

2. Errors from `CString::new` are silently ignored. Putting a NUL byte in what's supposed to be your function name should always be a programming error, so this function should panic on error instead.

3. The C-style prefix "namespacing" should be replaced with a proper Rust module, i.e. `nativescript::utils::profiling::add_data` (or just `nativescript::profiling::add_data`).

If you need help with the code changes, please feel free to join us on our development Discord: https://discord.gg/FNudpBD

bluenote10 commented:
> I understand that this is your first code contribution to this repo, and don't mean to sound harsh in any way

Not at all, always happy to learn, that's what I meant by "naive" ;). I've spend a few more thoughts now.

In general it would be good to expose both a low-level interface with zero overhead (e.g. I wanted to avoid imposing the overhead of `format!` unconditionally), combined with a convenient high level interface. This enables users to go for the low level interface directly if they aren't happy with the high level one. I've made this separation more clear now, by making the low level interface properly zero overhead, and referring users to the high level interface in the documentation.

For the high level interface I've considered three options:

1. A scope based profiling macro like
    ```rust
    profile!("foo", {
        // ...
    }
    ```

2. An explicit two line profiling macro like
    ```rust
    let p = profile!("foo");
    // ...
    p.measure();
    ```

3. A drop based profling macro like
    ```rust
    profile!(p, "foo");
    // ...
    // measures automatically until `p` gets dropped
    ```

Personally I find (1) is the cleanest option, but it also is very intrusive, because it adds two lines and an indentation level. I have experimented with it in my code (involving some nested measurements), and it was so intrusive that I cannot imagine keeping these `profile!` block around (just from a syntactical perspective, assuming there is mechanism that disables profiling by conditional compilation). Similarly (2) also still feels a bit too much. This leaves the drop based solution (3). This is a bit more tricky because one has to be conscious that measurement is determined by the drop, but for me it is minimal enough to tolerate them in my code.

This leaves question of how to control the conditional compilation of profiling. I think it doesn't make much sense to couple it to debug builds, because (1) profiling is not really meaningful in debug builds, and (2) switching from debug to release is not the most convenient thing within Godot, because the target output paths differ and typically it requires editing resource paths in the `.gdnlib` file. This suggests to use a feature gate instead, and the question becomes whether it should be opt-in or opt-out. I went for opt-out, because it makes it more easy for beginners to get going with profiling, and disabling the profiling is only really needed very late in the development process when shipping the final build.

bluenote10 commented:
I've spend a while actually doing some profiling on my app to get an impression of the user experience. I hope I can better explain now why I prefer solution (2) or (3) from a user perspective over a scope based solution. I see two points:

1. A frequent use case is to profile only a certain chunk of code, and with a scope based approach it is more cumbersome to "quickly" add/remove profiling instructions. For instance, say we want to measure just these two statements in isolation:

    ```rust
    let prev = compute_prev_element(...);
    let next = compute_next_element(...);
    ```
    With scope-based profiling this requires rewriting the code to
    ```rust
    let (prev, next) = profile!("compute_prev_next", || {
        (
            compute_prev_element(...),
            compute_next_element(...),
        )
    });
    ```
    whereas (2) or (3) doesn't require structural changes, and can be added/removed more easily, e.g.:
    ```rust
    profile!(profiler, "compute_prev_next");
    let prev = compute_prev_element(...);
    let next = compute_next_element(...);
    drop(profiler)
    ```

2. Like in the example, logical blocks (like functions and scopes) can sometimes be different from meaningful measurement units. Profiling instructions are of least importance, because they are even optimized away in the end. Therefore, they should only play a subordinate role. In my opinion, they should not dictate logical blocks or program flow.

Another open point was:

> I'd really prefer a procedural macro that doesn't require users to manually provide a string literal

From a practical perspective this is a quite handy feature to give the profiler identifiers more meaning. For instance it has helped to disambiguate all my `_process` by actually using `MyType._process`. Also note that in contrast to GDScript profiling, native profiler data cannot distinguish between "self" and "inclusive" time. To work-around that it has helped to temporarily place lots of profiler lines using a tag naming convention like this:
```rust
fn process(...) {
    profile!(p, "MyType._process");

    if some_condition {
        profile!(p, "MyType._process/preprocessing_case_1");    
        // ...
    } else {
        profile!(p, "MyType._process/preprocessing_case_2");    
        // ...
    }

    for elements in &self.elements {
        profile!(p, "MyType._process/element_iteration");
        // ...
    }

    profile!(p, "MyType._process/post_process");
    // ...
}
```
The naming convention then helps to identify the blocks in the profiler, and comparing the graphs gives an insight on time distribution:
![image](https://user-images.githubusercontent.com/3620703/91345964-792e7600-e7e0-11ea-9c7d-07f658b6a47b.png)
One may even binary-search with the `profile!` statements within a function body to identify bottlenecks.

Tomorrow I'll see what I can do about the open points.

toasteater commented:
> logical blocks (like functions and scopes) can sometimes be different from meaningful measurement units. Profiling instructions are of least importance, because they are even optimized away in the end. Therefore, they should only play a subordinate role. In my opinion, they should not dictate logical blocks or program flow.

But a drop-based API still dictate logical blocks and program flow, because owned values are dropped at the end of scope, and scope is tied to logical blocks. If the `add_data` function is exposed for people who want to measure time any way they want, I don't see why the high-level helper must try to cover every possible use case (and fail).

> disambiguate all my `_process` by actually using `MyType._process`

An attribute macro would still allow you to specify your own name as an option. It should also be possible to automatically put the module path in the identifier by invoking `std::module_path`. 

>  in contrast to GDScript profiling, native profiler data cannot distinguish between "self" and "inclusive" time.

Use a native profiler. They'll give you "self" and "inclusive" times and more and you don't have to even do any of this.

> To work-around that it has helped to temporarily place lots of profiler lines using a tag naming convention like this:

Having to do this usually indicates that your function is too big, and the profiled sections should be separated into smaller functions.

bluenote10 commented:
> But a drop-based API still dictate logical blocks and program flow, because owned values are dropped at the end of scope, and scope is tied to logical blocks.

A drop-based API does not dictate logical blocks in the sense that a desire to measure a certain group of lines requires to enclose that group in its own scope. It rather _allows to follow_ existing logical scopes where appropriate, with an escape hatch for marking a measuring range manually without the need to introduce a scope. When I was instrumenting my app the desirable measuring units coincided with existing logical scopes in roughly 80% of the cases. In these cases a scope based profiling macro would just introduce a scope where there is already a scope -- following up on the example above something like:

```rust
fn process(...) {
    profile!("MyType._process", || {
        if some_condition {
            profile!("MyType._process/preprocessing_case_1", || {
                // ...
            });
        } else {
            profile!("MyType._process/preprocessing_case_2", || {
                // ...
            });
        }

        for elements in &self.elements {
            profile!("MyType._process/element_iteration", || {
                // ...
            });
        }

        profile!("MyType._process/post_process", || {
            // ...
        });
    });
}
```
In the 20% cases where my desired measuring units did not coincide with an already existing scope (i.e. where I placed a manual `drop(profiler)`), a scope-based macro would also have made life more difficult, because these cases typically involved groups of `let` bindings that which would require the tedious tuple packing/unpacking refactoring mentioned before.

> Having to do this usually indicates that your function is too big, and the profiled sections should be separated into smaller functions.

That is a bit of an unfair assumption. I just checked: The longest function I instrumented was 15 lines. And being able to measure narrow ranges turned out to be useful even for very short functions. I should mention that my use case may be a bit special: My Godot bindings typically call directly into non-Godot-dependent libraries (e.g. I'm working on an audio engine library, completely independent of Godot). Therefore, I cannot move the Godot profiling statements any "deeper" into these libraries. Perhaps this explains why this "tag-based" profiling approach was preferred in my use case. (Of course this again calls for a native profiler, which I also tried, but regarding frame drops Godot's built-in profile was more insightful so far.)

In any case I have removed the higher level API now, and implemented an interface along what you had suggested. As a compromise, I hope it is okay to expose the low-level binding function directly giving users maximum flexibility.

toasteater reviewed (approved):
Looks good to me. Thanks!

bors r+
